File: leetcode/Combination_Sum.cpp
class Solution {
public:
    vector<vector<int> > combinationSum(vector<int> &candidates, int target) {
        if(candidates.empty()) return {};
        sort(candidates.begin(), candidates.end());
        
        vector<vector<int>> result;
        vector<int> sofar;
        comb_sum_helper(result, sofar, 0, target, candidates);
        
        return result;
    }
private:
    void comb_sum_helper(vector<vector<int>> &result, vector<int> &sofar, int sum, int target, vector<int> &remain){
        if(sum == target){
            result.push_back(sofar);
            return;
        }
        
        if(remain.empty()) return;
        
        int t = remain[0];
        remain.erase(remain.begin());
        
        comb_sum_helper(result, sofar, sum, target, remain);
        
        int i = 1;
        while(sum + i*t <= target){
            sofar.push_back(t);
            comb_sum_helper(result, sofar, sum + i*t, target, remain);    
            i++;
        }
        
        //recover from options
        while(i > 1) {
            sofar.pop_back();
            i--;
        }
        remain.insert(remain.begin(), t);
    }
};

File: leetcode/java/CombinationSum.java
// also check the coin change problem

[1, 2, 3, 6, 7] target = 7
// backtracking like we do coin change problem
//             cs(7)
               1    2    3   6   7   <-- this branch will use all elems instead of all combinations of one specific elem
//            cs(6)
              1
//          cs(5)

// Another to way to backtracking
//             cs(7)
         0*1     1*1    2*1    3*1  <-- this branch will use all the possible combinations of 1
//     cs(7)     cs(6)
        0*2      0*2                <-- use all combinations of 2
// More concise version
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        if(candidates == null || candidates.length == 0) return Collections.emptyList();

        List<List<Integer>> result = new ArrayList<>();
        csh(result, new ArrayList<>(), candidates, 0, target);
        return result;
    }
    
    private void csh(List<List<Integer>> result, List<Integer> sofar, int[] nums, int d, int t){
        int sum = sofar.stream().reduce(0, Integer::sum);
        if (sum == t) {
            result.add(new ArrayList<>(sofar));
            return;
        }
        if (d == nums.length) return;
        
        int cnt = 0;
        while(sum + cnt*nums[d] <= t) {
            csh(result, sofar, nums, d+1, t);
            sofar.add(nums[d]);
            cnt++;
        }
        for(int i = 0; i < cnt; i++) sofar.remove(sofar.size()-1);
    }
}
